features/dd-sdk-android-flags/src/main/kotlin/com/datadog/android/flags/featureflags/internal/repository/DefaultFlagsRepository.kt

```diff
@@ -8,36 +8,43 @@ package com.datadog.android.flags.featureflags.internal.repository
 
 import com.datadog.android.api.InternalLogger
 import com.datadog.android.api.feature.FeatureSdkCore
+import com.datadog.android.api.storage.datastore.DataStoreHandler
 import com.datadog.android.flags.featureflags.internal.model.PrecomputedFlag
+import com.datadog.android.flags.featureflags.internal.persistence.FlagsPersistenceManager
 import com.datadog.android.flags.featureflags.model.EvaluationContext
 import java.util.concurrent.atomic.AtomicReference
 
 internal class DefaultFlagsRepository(
     private val featureSdkCore: FeatureSdkCore,
+    private val instanceName: String,
+    private val dataStore: DataStoreHandler,
     private val internalLogger: InternalLogger = featureSdkCore.internalLogger
 ) : FlagsRepository {
-    // Atomic state - ensures context and flags are always consistent
     private data class FlagsState(val context: EvaluationContext, val flags: Map<String, PrecomputedFlag>)
     private val atomicState = AtomicReference<FlagsState?>(null)
 
+    private val persistenceManager = FlagsPersistenceManager(
+        dataStore = dataStore,
+        instanceName = instanceName,
+        internalLogger = internalLogger
+    ) { persistedState ->
+        persistedState?.let {
+            val loadedState = FlagsState(it.evaluationContext, it.flags)
+            atomicState.compareAndSet(null, loadedState)
+        }
+    }
+
     override fun setFlagsAndContext(context: EvaluationContext, flags: Map<String, PrecomputedFlag>) {
         val newState = FlagsState(context, flags)
         atomicState.set(newState)
-
-        internalLogger.log(
-            InternalLogger.Level.DEBUG,
-            InternalLogger.Target.MAINTAINER,
-            { "Set flags and context: ${flags.size} flags for context: ${context.targetingKey}" }
-        )
+        persistenceManager.saveFlagsState(context, flags)
```

Comment:
if save operation failed by some reason, is there any serious impact of having that value still in `atomicState` only?

Reply:
there won't be persistence between sessions, but we need to store the state in order to continue serving flags. I'll add in the datastore callback here so that we can log an error